nemo/collections/asr/parts/utils/vad_utils.py

```diff
@@ -325,41 +326,197 @@ def generate_vad_segment_table_per_file(pred_filepath, per_args):
+
+    Args:
+        per_args:
+            onset (float): onset threshold for detecting the beginning and end of a speech 
+            offset (float): offset threshold for detecting the end of a speech. 
+            pad_onset (float): adding durations before each speech segment
+            pad_offset (float): adding durations after each speech segment;
+            shift_len (float): amount of shift of window for generating the frame.
+    
+    Returns:
+        speech_segments(set): Set of speech segment in (start, end) format. 
+    """
+    shift_len = per_args.get('shift_len', 0.01)
+
+    onset = per_args.get('onset', 0.5)
+    offset = per_args.get('offset', 0.5)
+    pad_onset = per_args.get('pad_onset', 0.0)
+    pad_offset = per_args.get('pad_offset', 0.0)
+
+    onset, offset = cal_vad_onset_offset(per_args.get('scale', 'absolute'), onset, offset)
+
+    speech = False
+    speech_segments = set()  # {(start1, end1), (start2, end2)}
+    for i in range(1, len(sequence)):
+        # Current frame is speech
+        if speech:
+            # Switch from speech to non-speech
+            if sequence[i] < offset:
+                if start - pad_onset >= 0:
+                    speech_segments.add((start - pad_onset, i * shift_len + pad_offset))
+                else:
+                    speech_segments.add((0, i * shift_len + pad_offset))
+                start = i * shift_len
+                speech = False
+
+        # Current frame is non-speech
+        else:
+            # Switch from non-speech to speech
+            if sequence[i] > onset:
+                start = i * shift_len
+                speech = True
+
+    # if it's speech at the end, add final segment
+    if speech:
+        speech_segments.add((start - pad_onset, i * shift_len + pad_offset))
+
+    # Merge the overlapped speech segments due to padding
+    speech_segments = merge_overlap_segment(speech_segments)  # not sorted
+
+    return speech_segments
+
+
+def filtering(speech_segments, per_args):
+    """
+    Binarize predictions to speech and non-speech
+
+    Reference
+    Paper: Gregory Gelly and Jean-Luc Gauvain. "Minimum Word Error Training of RNN-based Voice Activity Detection", InterSpeech 2015. 
+    Implementation: https://github.com/pyannote/pyannote-audio/blob/master/pyannote/audio/utils/signal.py 
+    Args:
+        per_args:
+            min_duration_on (float): threshold for small non_speech deletion
+            min_duration_off (float): threshold for short speech segment deletion
+            filter_speech_first (boolean): Whether to perform short speech segment deletion first.
+
+    Returns:
+        speech_segments(set): Filtered set of speech segment in (start, end) format.  {(start1, end1), (start2, end2)}
+    """
+    min_duration_on = per_args.get('min_duration_on', 0.0)
+    min_duration_off = per_args.get('min_duration_off', 0.0)
+    filter_speech_first = per_args.get('filter_speech_first', True)
```

Comment:
two different defaults, in yaml its False and here its True

Reply:
Here is the simplest and conventional method. In yaml, the param is tuned with CH109